lib/util/propTypes.js

```diff
@@ -0,0 +1,671 @@
+'use strict';
+
+const annotations = require('./annotations');
+const propsUtil = require('./props');
+const variableUtil = require('./variable');
+const versionUtil = require('./version');
+
+/**
+ * Checks if we are declaring a props as a generic type in a flow-annotated class.
+ *
+ * @param {ASTNode} node  the AST node being checked.
+ * @returns {Boolean} True if the node is a class with generic prop types, false if not.
+ */
+function isSuperTypeParameterPropsDeclaration(node) {
+  if (node && (node.type === 'ClassDeclaration' || node.type === 'ClassExpression')) {
+    if (node.superTypeParameters && node.superTypeParameters.params.length > 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * Removes quotes from around an identifier.
+ * @param {string} the identifier to strip
+ */
+function stripQuotes(string) {
+  return string.replace(/^\'|\'$/g, '');
+}
+
+/**
+ * Retrieve the name of a key node
+ * @param {ASTNode} node The AST node with the key.
+ * @return {string} the name of the key
+ */
+function getKeyValue(context, node) {
+  if (node.type === 'ObjectTypeProperty') {
+    const tokens = context.getFirstTokens(node, 2);
+    return (tokens[0].value === '+' || tokens[0].value === '-'
+      ? tokens[1].value
+      : stripQuotes(tokens[0].value)
+    );
+  }
+  const key = node.key || node.argument;
+  return key.type === 'Identifier' ? key.name : key.value;
+}
+
+/**
+ * Iterates through a properties node, like a customized forEach.
+ * @param {Object[]} properties Array of properties to iterate.
+ * @param {Function} fn Function to call on each property, receives property key
+    and property value. (key, value) => void
+ */
+function iterateProperties(context, properties, fn) {
+  if (properties && properties.length && typeof fn === 'function') {
+    for (let i = 0, j = properties.length; i < j; i++) {
+      const node = properties[i];
+      const key = getKeyValue(context, node);
+
+      const value = node.value;
+      fn(key, value);
+    }
+  }
+}
+
+module.exports = {
+  instructions(context, components, utils) {
+    let stack = null;
+
+    // Used to track the type annotations in scope.
+    // Necessary because babel's scopes do not track type annotations.
+    const classExpressions = [];
+    const defaults = {skipShapeProps: true, customValidators: []};
+    const configuration = Object.assign({}, defaults, context.options[0] || {});
+    const customValidators = configuration.customValidators;
```

Comment:
These two config options come from two different rules. Put them here for now, maybe will need to consider if all rules using this functionality should take these as options:

Reply:
Actually I was partly wrong, `customValidators` is present in both `prop-types` and `no-unused-prop-types`. `skipShapeProps` comes from `no-unused-prop-types` only, because it can be hard to figure out that a prop access corresponds to a shape. Not sure it's needed in `prop-types`, will take a look.

One problem right now is `skipShapeProps` is `true` by default in `no-unused-prop-types`, whereas `prop-types` never ignores shape props. Will try to fix that by always building declaredPropTypes for shapes and checking the option when validating props.